update.go

```diff
@@ -512,3 +524,261 @@ func rollbackUpdatedFiles(currentApplicationDir string, updateFiles []updateFile
 	}
 	return
 }
+
+const ZipArchiveExtension = ".zip"
+
+var TarGzArchiveExtensions = []string{".tgz", ".tar.gz"}
+
+type versionFilesLoader struct {
+	version                Version
+	updateConfig           *UpdateConfig
+	getReplacementFileInfo func(loadedFilename string) (ReplacementFile, error)
+	destDir                string
+	createOrdinaryFiles    bool // create temp files otherwise USE CAREFULLY files could replace each other in ordinary naming
+}
+
+func (vfl versionFilesLoader) loadVersionFiles() ([]updateFile, error) {
+	assetsFilenames := vfl.version.getAssetsFilenames()
+	archivesFilenames := make([]string, 0)
+	for i, filename := range assetsFilenames {
+		isArchive := false
+		for _, ext := range append(TarGzArchiveExtensions, ZipArchiveExtension) {
+			if strings.HasSuffix(filename, ext) {
+				isArchive = true
+				break
+			}
+		}
+		if isArchive {
+			archivesFilenames = append(archivesFilenames, filename)
+			continue
+		}
+		assetsFilenames[i-len(archivesFilenames)] = filename
+	}
+	assetsFilenames = assetsFilenames[:len(assetsFilenames)-len(archivesFilenames)]
+
+	updateFilesInfo, err := vfl.loadUpdateFilesFromSource(assetsFilenames)
+	if err != nil {
+		return nil, err
+	}
+	archivesUpdateFilesInfo, err := vfl.loadUpdateFilesFromSourceArchives(archivesFilenames)
+	if err != nil {
+		return nil, err
+	}
+	return append(updateFilesInfo, archivesUpdateFilesInfo...), nil
+}
+
+func (vfl versionFilesLoader) loadUpdateFilesFromSourceArchives(archivesFilenames []string) ([]updateFile, error) {
+	updateFilesInfo := make([]updateFile, 0)
+	for _, archive := range archivesFilenames {
+		fExt := filepath.Ext(archive)
+		fPath, err := vfl.loadUpdateFileFromSource(archive)
+		if err != nil {
+			return nil, err
+		}
+		var ufi []updateFile
+		if fExt == ZipArchiveExtension {
+			ufi, err = vfl.unpackZipArchive(fPath)
+
+		} else {
+			ufi, err = vfl.unpackTarGzArchive(fPath)
+		}
+		if err != nil {
+			return nil, err
+		}
+		updateFilesInfo = append(updateFilesInfo, ufi...)
+	}
+	return updateFilesInfo, nil
+}
+
+func (vfl versionFilesLoader) unpackZipArchive(archiveFPath string) ([]updateFile, error) {
+	updateFilesInfo := make([]updateFile, 0)
+	zR, err := zip.OpenReader(archiveFPath)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		zRCloseErr := zR.Close()
+		if err != nil && zRCloseErr != nil {
+			err = fmt.Errorf("%v; zip close error: %v", err, zRCloseErr)
+		}
+		if err == nil {
+			err = zRCloseErr
+		}
+	}()
+
+	for _, file := range zR.File {
+		if file.FileInfo().IsDir() {
+			continue
+		}
+		fName := filepath.Base(file.Name)
+		replacementFileInfo, err1 := vfl.getReplacementFileInfo(fName)
+		if err1 != nil {
+			err = err1
+			return nil, err
```

Comment:
I think you can simply return `err1`, without assigning it like `err = err1`, it'll be the same

Reply:
But how defer set close err if I do this way?

Reply:
Oh, i tested it, and all my defer sentence isn't work(

Reply:
![image](https://user-images.githubusercontent.com/42908925/150868872-1a14305d-f7d1-4b93-8ab5-649386e6beb7.png)
It works fine if I do this way, but it's not clear for me, because of variable reassignment. So As I understand, when we declare variable with function declaration we have access to it beyond the function, but, when I reassign it it's a new var in function scope, isn't it? Why I still have opportunity to change it(((

Reply:
WTF, and still works fine if I do this way 
![image](https://user-images.githubusercontent.com/42908925/150869894-cbe6ba2f-d83c-4057-b848-3ff8970ca268.png)

Reply:
But doesn't work this way 
![image](https://user-images.githubusercontent.com/42908925/150870047-b0c493e0-43ff-474f-9c0b-b809cc5bb0f0.png)

Reply:
Feel sad

Reply:
KK, reading this article gave me some answers https://stackoverflow.com/questions/37248898/how-does-defer-and-named-return-value-work

But, i still don't understand why is last variant didn't do what i whant)

Reply:
heh, by design 
https://stackoverflow.com/questions/48680222/what-is-the-difference-between-named-return-value-and-normal-return-value

Reply:
Ok, i will fix that, thanks, it was intresting investigation

Reply:
Nice catch with missing `err` named return variable declaration!

As for the issue

> but, when I reassign it it's a new var in function scope, isn't it?

No, this is the same scope (one function), so you are reusing variable and reassigning it 

> But, i still don't understand why is last variant didn't do what i whant)

Because you don't use named return? It only works with defer + named return

Reply:
Really appreciate your comments and explanations! Thank you! Now it's clear for me, I do some update, to fix my misunderstanding of errors wrapping in defer